File: ui/src/api/client-account.js
/** @import { Account, UpdateFields } from './types.js' */
import { SystemControllerClient } from './core.js'

/**
 * Create a new user account.
 * @param {string} username - Unique username.
 * @param {string} password - Password (must be at least 8 characters).
 * @param {string} email - Email address (validated against standard format).
 * @param {string} phone - Phone number (digits with optional formatting).
 * @param {string} realName - Display name.
 * @param {boolean} admin - When true, the account receives administrator privileges. Ignored when `wireguard` is true.
 * @param {boolean} [wireguard] - When true, create a WireGuard-only account scoped to `networks` instead of a normal account.
 * @param {string[]} [networks] - Networks the WireGuard-only account may enroll peers on. Must be non-empty when `wireguard` is true.
 * @returns {Promise<Account>}
 */
SystemControllerClient.prototype.createAccount = async function (username, password, email, phone, realName, admin, wireguard = false, networks = []) {
  return this.postJSON('/account/create', {
    username,
    password,
    email,
    phone,
    real_name: realName,
    admin,
    wireguard,
    networks,
  })
}

/**
 * @param {string} username
 * @returns {Promise<Account>}
 */
SystemControllerClient.prototype.getAccount = async function (username) {
  return this.postJSON('/account', { username })
}

/**
 * Update fields on an existing account. Only provided fields are changed.
 * @param {string} username - Account to update.
 * @param {UpdateFields} fields - Fields to change (password, email, phone, real_name, admin).
 * @returns {Promise<Account>}
 */
SystemControllerClient.prototype.updateAccount = async function (username, fields) {
  return this.postJSON('/account/update', { username, fields })
}

/**
 * @param {string} [sortBy]
 * @param {string} [sortOrder]
 * @param {number} [limit]
 * @param {number} [offset]
 * @param {string} [search]
 * @returns {Promise<{entries: Account[], has_more: boolean, total_pages: number, total_count: number}>}
 */
SystemControllerClient.prototype.listAccounts = async function (sortBy, sortOrder, limit, offset, search) {
  const params = new URLSearchParams()
  if (sortBy) params.set('sort_by', sortBy)
  if (sortOrder) params.set('sort_order', sortOrder)
  if (limit) params.set('limit', String(limit))
  if (offset) params.set('offset', String(offset))
  if (search) params.set('search', search)
  const qs = params.toString()
  return this.getJSON(`/account${qs ? `?${qs}` : ''}`)
}

/**
 * @param {string} username
 * @returns {Promise<void>}
 */
SystemControllerClient.prototype.disableAccount = async function (username) {
  await this.post('/account/disable', { username })
}

/**
 * @param {string} username
 * @returns {Promise<void>}
 */
SystemControllerClient.prototype.enableAccount = async function (username) {
  await this.post('/account/enable', { username })
}

File: src/svc/systemcontroller/client_account.go
package systemcontroller

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"

	"gitea.com/town-os/town-os/src/account"
)

// --- Account ---

// CreateAccount creates a new user account with the given profile fields.
// When admin is true the account receives administrator privileges.
func (c *SystemdClient) CreateAccount(ctx context.Context, username, password, email, phone, realName string, admin bool) (_ *account.Account, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, CreateAccountRequest{
		Username: username, Password: password,
		Email: email, Phone: phone, RealName: realName, Admin: admin,
	})

	resp, err := c.postJSON(ctx, "account/create", pr)
	if err != nil {
		return nil, fmt.Errorf("%w: CreateAccount: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "POST", "account/create")
	}

	var acct account.Account
	return &acct, json.NewDecoder(resp.Body).Decode(&acct)
}

// CreateGrantedAccount creates a non-admin account holding grants, scoped to
// networks. It may reach only the routes those grants unlock, on those networks.
func (c *SystemdClient) CreateGrantedAccount(ctx context.Context, username, password, email, phone, realName string, grants, networks []string) (_ *account.Account, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, CreateAccountRequest{
		Username: username, Password: password,
		Email: email, Phone: phone, RealName: realName,
		Grants: grants, Networks: networks,
	})

	resp, err := c.postJSON(ctx, "account/create", pr)
	if err != nil {
		return nil, fmt.Errorf("%w: CreateGrantedAccount: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "POST", "account/create")
	}

	var acct account.Account
	return &acct, json.NewDecoder(resp.Body).Decode(&acct)
}

// GetAccount retrieves a user account by username.
func (c *SystemdClient) GetAccount(ctx context.Context, username string) (_ *account.Account, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, GetAccountRequest{Username: username})

	resp, err := c.postJSON(ctx, "account", pr)
	if err != nil {
		return nil, fmt.Errorf("%w: GetAccount: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "POST", "account")
	}

	var acct account.Account
	return &acct, json.NewDecoder(resp.Body).Decode(&acct)
}

// UpdateAccount modifies fields on an existing user account. Only non-nil
// fields in the [account.UpdateFields] struct are applied.
func (c *SystemdClient) UpdateAccount(ctx context.Context, username string, fields account.UpdateFields) (_ *account.Account, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, UpdateAccountRequest{Username: username, Fields: fields})

	resp, err := c.postJSON(ctx, "account/update", pr)
	if err != nil {
		return nil, fmt.Errorf("%w: UpdateAccount: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "POST", "account/update")
	}

	var acct account.Account
	return &acct, json.NewDecoder(resp.Body).Decode(&acct)
}

// DisableAccount prevents the named user from authenticating.
func (c *SystemdClient) DisableAccount(ctx context.Context, username string) error {
	pr, pw := io.Pipe()
	go pipeEncode(pw, DisableAccountRequest{Username: username})

	return c.postClient(ctx, "account/disable", pr)
}

// EnableAccount re-enables a previously disabled user account.
func (c *SystemdClient) EnableAccount(ctx context.Context, username string) error {
	pr, pw := io.Pipe()
	go pipeEncode(pw, EnableAccountRequest{Username: username})

	return c.postClient(ctx, "account/enable", pr)
}

// ListAccounts returns a paginated list of all user accounts.
func (c *SystemdClient) ListAccounts(ctx context.Context, params ListParams) (_ *PageResult[account.Account], err error) {
	resp, err := c.getClient(ctx, "account"+params.QueryString())
	if err != nil {
		return nil, fmt.Errorf("%w: ListAccounts: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "GET", "account")
	}

	var page PageResult[account.Account]
	return &page, json.NewDecoder(resp.Body).Decode(&page)
}
